### ISSUE TITLE
Audio Support

### ISSUE THREAD

**Issue by TomMD** [acowley/ffmpeg-light#21]
It might require more dependencies, but supporting audio seems like an obvious next step.  I don't have any specifics here, such as a pre-existing audio frame type already available on Hackage, so this requires quite a bit of fleshing out.

**Comment by acowley** [comment]
I did [play around](https://github.com/acowley/EasyAudio) with SDL2 audio before, and it was relatively straightforward. In the absence of any other feelings, I'd give that a shot again. I haven't yet looked at how ffmpeg bubbles up the audio data, though.

If we go that way, we would [provide a callback to SDL](http://hackage.haskell.org/package/sdl2-2.1.1/docs/SDL-Audio.html#t:OpenDeviceSpec) that would shuffle the data from ffmpeg into an SDL buffer.

**Comment by leftaroundabout** [comment]
Why the need for anything like SDL? Couldn't we just have something quite like [imageReader](http://hackage.haskell.org/package/ffmpeg-light-0.11.1/docs/Codec-FFmpeg.html#v:imageReader), except that it would load chunks of audio data instead of `Image`s?

**Comment by acowley** [comment]
Yeah, that's the right way to do it in the core library. An example executable showing SDL integration is probably enough to show users a sample complete pipeline.

**Comment by DiegoNolan** [comment]
I've started working on this but boy is this library difficult to work with. You can see my progress here.

https://github.com/DiegoNolan/ffmpeg-light/commit/f3c3f525cc48ca83b16a78058d931e4a9da2a978

There are some things I want out of this. I want to be able to have an audio reader that gives out the data in chunks. I want to be able to resample audio to combine them and overlay them over videos or extract the audio from the video. I want this as a cleaner abstraction that what FFmpeg provides in the C interface which is quite cumbersome.

I'm not sure how we should represent the audio. FFmpeg has AVSampleFormat. They represent audio in unsigned and signed 8, 16, and 32 bit arrays as well as floats and doubles. It seems like it would be better to have a common format that different audio encodings could use to be used together. The presence of channels complicates things quite a bit since it is essentially multiple vectors of audio.

Just wanted to see your feedback on API design and the progress I've made so far. We could also look at SFML which provides a way to read audio but it doesn't have as much codec support as FFmpeg. It also may provide a newer perspective when compared to SDL.

I also think ideally there should be a pipes interface for all of this to really make it user friendly and usable but this is a lot more work down the line.

**Comment by leftaroundabout** [comment]
Kudos for all that hard work!

I think an FFmpeg library should keep as close to what FFmpeg spits out as possible while being still _safe_ to use, and not going completely against the grain of the language.

Sure, for end applications we'd eventually want higher-level interfaces – I personally hope to some day get [timed-media](https://github.com/leftaroundabout/timed-media) ready, with _no such notion_ as sample rate etc. (working completely in frequency space). But that sort of stuff has nothing to do with FFmpeg.

**Comment by DiegoNolan** [comment]
Hmm. I almost have it working but I have a bug. Well I extract audio and write it to a mp3 or something there is a bunch of distortion and it doubles the length. Can't seem to figure out where the bug is.

**Comment by leftaroundabout** [comment]
Blind guess: “_bunch of distortion and doubles the length_” sounds like stereo-interlacing gone awry.

**Comment by DiegoNolan** [comment]
I also get a seg fault with another video file. If I do ffprode on each of them they are mono. I don't try resampling either. Just going from mp4 with an aac audio stream to a simple aac.

**Comment by DiegoNolan** [comment]
Well I figured out the seg fault issue but not the other.

**Comment by DiegoNolan** [comment]
Okay, I was able to extract the audio from a video but I had to remove the resampling part. Also I'm unsure how to represent the frame data. I was only able to do it by passing the decoding frames directly to be re encoding.

https://ffmpeg.org/doxygen/3.1/swresample_8c.html#a39cb25282265493348bb8604120b87b9

swr_convert seems to take an array of arrays of fixed size. AVFrame does something similar. I'm guessing each array represents a channel of audio.
